mackerel-plugin-accesslog/lib/accesslog.go

```diff
@@ -0,0 +1,183 @@
+package mpaccesslog
+
+import (
+	"bufio"
+	"flag"
+	"fmt"
+	"io"
+	"log"
+	"os"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/Songmu/axslogparser"
+	"github.com/Songmu/postailer"
+	mp "github.com/mackerelio/go-mackerel-plugin"
+	"github.com/mackerelio/golib/pluginutil"
+	"github.com/montanaflynn/stats"
+)
+
+// AccesslogPlugin mackerel plugin
+type AccesslogPlugin struct {
+	prefix    string
+	file      string
+	posFile   string
+	noPosFile bool
+}
+
+// MetricKeyPrefix interface for PluginWithPrefix
+func (p *AccesslogPlugin) MetricKeyPrefix() string {
+	if p.prefix == "" {
+		p.prefix = "accesslog"
+	}
+	return p.prefix
+}
+
+// GraphDefinition interface for mackerelplugin
+func (p *AccesslogPlugin) GraphDefinition() map[string]mp.Graphs {
+	labelPrefix := strings.Title(p.prefix)
+	return map[string]mp.Graphs{
+		"access_num": {
+			Label: labelPrefix + " Access Num",
+			Unit:  "integer",
+			Metrics: []mp.Metrics{
+				{Name: "total_count", Label: "Total Count"},
+				{Name: "5xx_count", Label: "HTTP 5xx Count", Stacked: true},
+				{Name: "4xx_count", Label: "HTTP 4xx Count", Stacked: true},
+				{Name: "3xx_count", Label: "HTTP 3xx Count", Stacked: true},
+				{Name: "2xx_count", Label: "HTTP 2xx Count", Stacked: true},
+			},
+		},
+		"access_rate": {
+			Label: labelPrefix + " Access Rate",
+			Unit:  "percentage",
+			Metrics: []mp.Metrics{
+				{Name: "5xx_percentage", Label: "HTTP 5xx Percentage", Stacked: true},
+				{Name: "4xx_percentage", Label: "HTTP 4xx Percentage", Stacked: true},
+				{Name: "3xx_percentage", Label: "HTTP 3xx Percentage", Stacked: true},
+				{Name: "2xx_percentage", Label: "HTTP 2xx Percentage", Stacked: true},
+			},
+		},
+		"latency": {
+			Label: labelPrefix + " Latency",
+			Unit:  "float",
+			Metrics: []mp.Metrics{
+				{Name: "99_percentile", Label: "99 Percentile"},
+				{Name: "95_percentile", Label: "95 Percentile"},
+				{Name: "90_percentile", Label: "90 Percentile"},
+				{Name: "average", Label: "Average"},
+			},
+		},
+	}
+}
+
+func (p *AccesslogPlugin) getPos() string {
+	base := p.file + ".pos.json"
+	if p.posFile != "" {
+		if filepath.IsAbs(p.posFile) {
+			return p.posFile
+		}
+		base = p.posFile
+	}
+	return filepath.Join(pluginutil.PluginWorkDir(), "mackerel-plugin-accesslog.d", base)
```

Comment:
The ".d" is added to clearly indicate that it is a directory, because `filepath.Join(pluginutil.PluginWorkDir(), "mackerel-plugin-accesslog")` conflicts with the file genareted by plugin to store the latest value (for `fetchLastValues`).